### PR TITLE
Reduce storage / billing costs for email alerts.

### PR DESCRIPTION
**Pull request by mikelehen**
This PR reduces the amount of data we will store long-term for past email alerts. We will:

* Add a `deleteAt` field to the docs written to Firestore that will trigger a TTL policy to delete alert docs after 2 weeks.
* Write stats after sending emails since we'll no longer be able to calculate them once the docs are deleted. We write per-snapshot stats to `/alert-stats/[snapshot]` and global stats to `/info/alert-stats`.
* I'm also checking in a couple scripts I'm using to generate historical stats and delete historical documents. These will only be run once, but I'm checking them in for future reference.

I've tested everything on the [covidactnow-dev Firestore instance](https://console.firebase.google.com/project/covidactnow-dev/firestore/data/) (except the fact that the TTL policy runs, since that will take 2+ weeks to kick in, but it should be very straightforward).  If you look there you'll see:

1. Historical stats have been generated (by my generate-historical-stats) script for all past snapshots, e.g.: image
2. The past `snapshots/<snapshot>` and `/vaccination-alerts/<snapshot>` locations are gone because I ran the `delete-historical-docs` script to delete them.
3. I did a fresh email alert run for snapshot 4388 and it marked the generated docs with a `deleteAt` value 2 weeks in the future: image
4. And it generated appropriate stats: 
image

I have also run the `generate-historical-stats.ts` scripts on staging and prod so they have `alert-stats` documents for each snapshot, and `info/alert-stats` global stats populated. (Fun fact, we've sent 13.7M email alerts to date)

Once this is reviewed / merged, I will go ahead and run the `delete-historical-docs.ts` script on staging and prod to clean up the past documents.

### CODE REVIEW

File: scripts/alert_emails/create_lists_to_email.ts [hunk 1]
```diff
@@ -54,11 +54,18 @@ async function updateSnapshotEmails(alertsFileName: string, snapshot: number) {
...
+          // future and we have a Firestore TTL policy configured to delete them.
+          const deleteAt = new Date();
+          deleteAt.setDate(deleteAt.getDate() + 14);
+
           try {
             await db
               .collection('snapshots')
               .doc(`${snapshot}/locations/${fips}/emails/${email}`)
-              .create({ sentAt: null });
+              .create({ sentAt: null, deleteAt });
```
Comment (mikelehen) [inline scripts/alert_emails/create_lists_to_email.ts:68]: This is really the main "fix" so that in the future the documents we create while sending email alerts will get deleted after 2 weeks.

File: scripts/alert_emails/misc/delete-historical-docs.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/**
+ * Script to go through all the historical sent emails in Firestore and delete
+ * them to save space.
+ */
```
Comment (mikelehen) [inline scripts/alert_emails/misc/delete-historical-docs.ts:4]: I checked in these scripts for future reference, but we should never need to run them again and they're not particularly high quality. 🤷

File: scripts/alert_emails/misc/generate-historical-stats.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
+/**
+ * Script to go through all the historical sent emails and generate aggregated
+ * stats about emails sent. Stats are incrementally stored to a local file (to
+ * support resuming progress in case an error is reported) and then stored to
+ * Firestore at the end.
+ */
```
Comment (mikelehen) [inline scripts/alert_emails/misc/generate-historical-stats.ts:6]: Ditto

File: scripts/alert_emails/misc/generate-historical-stats.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+import { getFirestore } from '../../common/firebase';
+import FirebaseFirestore from '@google-cloud/firestore';
+import path from 'path';
+import * as fs from 'fs';
+
+// File to store stats to. This is also used to support resuming in case there's an error.
+const STATS_FILE = path.join(__dirname, 'stats.json');
```
Comment (mikelehen) [inline scripts/alert_emails/misc/generate-historical-stats.ts:14]: FWIW- I built in this persistence / resume logic because in the past I've had trouble reading / processing all of our email alert docs on prod in one go... But this time when I ran it on prod it ran to the end with no problems, so maybe it wasn't really necessary. 🤷

File: scripts/alert_emails/readme.md [hunk 5]
```diff
@@ -42,12 +42,13 @@ The github action should complete in 20 minutes or so and list the number of ema
...
-2. Run `yarn create-lists-to-email fipsToAlertFilename <snapshot>` which will write to firestore `snapshots/<snapshot>/locations/<fips>/emails` for all the users that need to recieve emails for each of the snapshots/fips.
+2. Run `yarn create-lists-to-email alerts.json <snapshot>` which will write to firestore `snapshots/<snapshot>/locations/<fips>/emails` for all the users that need to recieve emails for each of the snapshots/fips.
 
 ### Sending the email
 
-1. If you have run the above steps and have users signed up for emails in the following locations you can run `yarn send-emails alerts.json 497 true <youremail>` which will send the emails to those users on dev.
-2. You can also passin an extra param `yarn send-emails alerts.json 497 true <youremail>` which will send all the emails to one email address for testing.
+1. If you have run the above steps and have users signed up for emails in the following locations you can run `yarn send-emails alerts.json <snapshot> true` which will send the emails to those users on dev.
+
+NOTE: You can pass `false` instead of `true` to do a dry-run instead: `yarn send-emails alerts.json <snapshot> false`.
```
Comment (mikelehen) [inline scripts/alert_emails/readme.md:51]: I made a few drive-by improvements to this doc as I was remembering how to use these scripts.  AFAICT these docs about passing "youremail" to the send-emails script were just wrong.  We have functionality like that for the vaccination alerts script, but not this one.

File: scripts/alert_emails/send_emails.ts [hunk 6]
```diff
@@ -7,6 +7,14 @@ import { generateAlertEmailData, readAlerts, resolveSnapshot } from './utils';
 
 const BATCH_SIZE = 20;
 
+interface SnapshotStats {
+  sent: Date;
+  locationsWithAlerts: number;
+  totalEmailsSent: number;
+  uniqueEmailAddresses: number;
+  invalidEmailsRemoved: number;
+}
```
Comment (mikelehen) [inline scripts/alert_emails/send_emails.ts:16]: This is the other part of the "fix" in that going forward we're going to collect stats after every email alert run so that we have per-snapshot stats and a global total of emails sent.

File: scripts/alert_emails/send_emails.ts [hunk 7]
```diff
@@ -78,7 +106,7 @@ async function setLastSnapshotNumber(
     return db
       .collection(`snapshots/${currentSnapshot}/locations/${fips}/emails/`)
       .doc(email)
-      .set({ sentAt: admin.firestore.FieldValue.serverTimestamp() });
+      .update({ sentAt: admin.firestore.FieldValue.serverTimestamp() });
```
Comment (mikelehen) [inline scripts/alert_emails/send_emails.ts:109]: Using `set()` would clobber the `deleteAt` field we wrote previously, so we use `update()` instead.

**Review by smcclure17 — APPROVED** [review]
Looks great to me! Thanks for taking this on, and for the walk through at standup 🙏
